rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,324 @@
+the WAL.
+
+SlateDB can improve over other implementations during the recovery process. In
+the implementations I tested, WAL recovery occurs asynchronously after the
+database is opened. As a result, the only way to warn the developer of WAL
+corruption is via a logged warning.
+
+To support operators who value durability and to ensure durability by default,
+I propose that `DB::open_with_opts()` perform WAL recovery synchronously. This
+method will return any WAL corruption found as errors and abort the database
+open process until the corruption is resolved. The error returned should be a
+distinct type that can easily be matched to determine if the error is due to
+non-transient corruption or a transient error (such as network connectivity
+issues).
+
+To support developers who value availability, we will introduce a new database
+option called `DbOptions.wal_auto_recovery`, similar in naming to
+`DbOptions.wal_enabled`. If set to `true`, then `DB::open_with_opts()` will
+perform WAL recovery asynchronously and automatically repair the WAL to the
+best of its ability, logging any corruption and repairs performed.
+
+To support the repair of the WAL, I propose introducing two new methods:
+`DB::wal_inspect()` and `DB::wal_repair()`. These methods will provide
+developers with tools to inspect the current contents of the WAL, including
+corrupted entries, and to repair the WAL to the best of its ability via entity
+recovery or truncation of the WAL.
+
+NOTE: `DbOptions.wal_auto_recovery` is a convenience for developers, as the
+same result can be achieved by calling `DB::open_with_opts()`, immediately
+calling `DB::wal_repair()`, and then calling `DB::open_with_opts()` again.
+
+### Handling SST Corruption
+Corruption of the SST can occur at any stage before, during, or after a
+database operation. Therefore, the most suitable place to report corruption is
+during the `get` call to retrieve keys from the database. If a key is NOT found
+and SlateDB detects corruption while searching for the key, an error should be
+reported. However, if SlateDB encounters corruption while searching for AND
+finding a key, it should not return an error. This ensures the database remains
+available even in the presence of corruption.
+
+The error returned should be a distinct type which can easily be matched to
+determine if the error is due to non transient corruption or a transient error.
+(network connectivity, etc..)
+
+The rationale behind this approach is rooted in the assumption that SlateDB's
+design is a manifestation of a physical system. Linux system administrators
+would not use physical disks that refuse access to all usable sectors due to a
+single bad sector. Such a disk would be impractical in many LVM (Logical Volume
+Manager) or RAID configurations. Similarly, the ability for developers who wish
+to use SlateDB to achieve Logical Durability is severely compromised if access
+to uncorrupted data blocks is not allowed.
```

Comment:
It would help me to clarify exactly what "availability" means here, and what "corrupt" means as well. I think there are a few different corruption states for an SST:

1. SST is WAL
2. SST isn't WAL
3. SST block is corrupt
4. SST block index is corrupt
5. SST block filter is corrupt

Then there is the state of a KV pair:

1. KV exists in a corrupt SST or block, but there is a more recent KV in a healthy (newer) SST.
2. KV is in a corrupt SST, but there is an older KV in a healthy (older) SST.
3. KV is in a corrupt SST and no other SST with the KV exists.

Then there is available/unavailable.

1. GETs return the most recent uncorrupt KV.
2. GETs return the most recent KV if it is uncorrupt. If it's corrupt in a corrupt block (3, above), an error is returned for the GET. GETs on KVs in uncorruot blocks continue as usual.
3. Any corrupt block results in the DB returning errors for ALL GETs.
4. Any corrupt SST indexes (4, above) results in an error return for all GETs.
5. Any corrupt SST (index, block, filter) results in errors for all GETs.

I have probably missed a few other states as well. But the point I'm getting at is that I think we can be partially available, which IMO might be the best default. I'm thinking:

- If a KV's most recent value is in a corrupt block error just that GET.
- If a KV's most recent value is in a non-corrupt block, it returns successfully l
- If an SST's index is corrupt and a GET has not yet found the KV in an earlier SST, and error is returned. This effectively prevents reads in SSTs older than the first corrupt SST. I think this is the right default behavior since we don't know if a KV is in the corrupt SST or not.
- We offer a single config (independent of WAL or compacted SST) that relaxes these constraints to skip all corrupt blocks and SSTs.